Optimize instance creation in __getattr__ [PREVIEW]

nicoe opened the pull request.

JCavallo reviewed (commented):
Si je comprends bien, le fait de se "ballader" dans le widget va trigger les on_change...

Ça me semble un peu violent, un utilisateur qui doit changer l'année / le mois risque de déclencher les on_change pas mal de fois. Ça ne serait pas possible de plutôt se brancher sur le focus-out ?

JCavallo reviewed (changes requested):
KO chez moi, le widget ne marche plus du tout

nicoe committed:
Optimize instance creation in __getattr__ [PREVIEW]

Accessing a M2O field from a X2M field could result in as many
instantiations as there are records in the X2M field. It could be a
performance bottleneck due to the many creations of Model kwargs and
potential useless calls of transaction.set_context().

Thus the optimization is twofold:

 - Create an additional code path when the M2O field modifies the
   context
 - Cache the kwargs creation process to prevent extra calls resulting in
   the same dictionary

nicoe committed:
Use read to fetch reverse field value of O2M field in get [PREVIEW]

Using a browse there in order to get the id is a performance bottleneck
as it will result in the instanciation of potentially thousands of
models.

nicoe committed:
Apply optimization to M2M

JCavallo committed:
Typo
